### PR TITLE
Add an LRU cache to avoid decrypting the same message more than once.

### PR DESCRIPTION
**Pull request by bemasc**

### CODE REVIEW

File: src/socketio.quiver.js [hunk 1]
```diff
@@ -1253,9 +1258,20 @@ QuiverSocialProvider.prototype.onEncryptedMessage_ = function(server, msg) {
...
 
-  // TODO: Memoize decryption to save CPU time for identical messages sent
-  // through different paths.
-  this.pgp_.verifyDecrypt(msg.cipherText, msg.key).then(function(plain) {
+  // Memoize decryption.  For security, it's important that msgKey be
+  // non-collidable even if the sender is hostile.
+  var bytes = new Uint8Array(msg.cipherText);
+  var msgKey = msg.key + ';' + Array.prototype.join.call(bytes, ':');
+
+  /** @type {Promise<!FreedomPgpDecryptResult>} */ var plainText;
```
Comment (dborkan) [inline src/socketio.quiver.js]: Can you call this something like getPlainText or plainTextPromise to emphasize its a promise and not a string?

File: externs/freedom_extern.js [hunk 2]
```diff
@@ -96,6 +96,14 @@ function FreedomCrypto() {};
  */
 FreedomCrypto.prototype.getRandomBytes = function(numBytes) {}
 
+/**
+ * @typedef {{
+ *   data: !ArrayBuffer,
+ *   signedBy: Array<string>
+ * }}
+ */
+var FreedomPgpDecryptResult;
```
Comment (dborkan) [inline externs/freedom_extern.js:105]: Why is this change needed?  Doesn't verifyDecrypt still return an array buffer?

**Comment by dborkan** [comment]
Just a few small comments on variable names and such, otherwise :+1:

File: src/socketio.quiver.js (hunk 1 above)
Comment (bemasc) [inline src/socketio.quiver.js]: Done

File: externs/freedom_extern.js (hunk 2 above)
Comment (bemasc) [inline externs/freedom_extern.js:105]: It's just for readability, because `FreedomPgpDecryptResult` is clearer than `{data: !ArrayBuffer, signedBy Array<string>}`, and it now appears in two places in `socketio.quiver.js`.
